How-to-upgrade-review-version.md

```diff
@@ -0,0 +1,399 @@
+
+You can
+
+1. Add missing types in dependencies.
+2. Fix references to types in dependencies.
+4. Add missing types in your own code.
+3. Work around missing types in dependencies.
+5. Fix errors in existing types.
+6. Add type annotations to everything else.
+
+## Add missing types in dependencies
+
+Let's start with @types/shelljs. In Makefile.js, I see a few errors.
+The first is that the module `require('shelljs/make')` isn't found.
+The second group of errors is that the names 'find', 'echo' and a few
+others aren't found.
+
+These errors are related. It turns out that shelljs doesn't even
+include shelljs/make right now. It's completely missing. I looked it
+up and shelljs/make does two things:
+
+1. Add a global object named 'target' that allows you to add make
+targets.
+2. Add the contents of the parent shelljs to the global scope.
+
+So I want to add this to Definitely Typed. My first step is to create
+make.d.ts inside node_modules/@types/shelljs/. This is just for
+development, because it makes it super easy to test that I'm actually
+fixing the missing stuff.
+
+I start with this:
+
+```ts
+import * as shelljs from './';
+declare global {
+  const cd: typeof shelljs.cd;
+  const pwd: typeof shelljs.pwd;
+  // ... all the rest ...
+}
+```
+
+Then I add the type for 'target' to the globals as well:
+
+```ts
+const target: {
+  all?: Target;
+  [s: string]: Target;
+}
+interface Target {
+  (...args: any[]): void;
+  result?: any;
+  done?: boolean;
+}
+```
+
+This exposes a couple more errors. See the section on fixing errors in
+existing types for how to fix those.
+
+Now we want to publish this to Definitely Typed:
+
+1. Fork DefinitelyTyped on github.
+2. `git clone https://github.com/your-name-here/DefinitelyTyped`
+3. `cp node_modules/@types/shelljs/make.d.ts ~/DefinitelyTyped/types/shelljs/`
+4. `git checkout -b add-shelljs-make`
+5. Commit the change and push it to your github fork.
+6. Create a PR for the change.
+
+If there are lint problems, the CI run on Travis will catch them.
+
+For more detail on writing definitions for Definitely Typed, [see the
+Declaration section of the Typescript handbook](http://www.typescriptlang.org/docs/handbook/declaration-files/introduction.html).
+
+## Fix references to types in dependencies
+
+In analyze-scope.js, I see quite a few errors about missing estree
+types like Identifier and ClassDeclaration. That's weird, because
+those types *do* exist in estree. The problem is that they're not
+imported. I'd like to write
+
+```js
+import { Identifier, ClassDeclaration } from "estree";
+```
+
+But this doesn't work because those are types, not values. The import
+will fail at runtime. Instead, you need to use an *import type*. An
+import type is just like a dynamic import, except that it's used as a
+type. So, just like you could write:
+
+```js
+const fs = import("fs");
+```
+
+to dynamically import `fs`, you can write:
+
+```ts
+var id: import("estree").Identifier = ...
```

Comment:
But I can't add this to my code, can I? These are JS files

Reply:
Ah, good point. I'll stick with JS.